### PR TITLE
expected(HW04) .to eql(performed)

### PR DESCRIPTION
**Pull request by kjkkuk**
from Dima :)

### CODE REVIEW

File: HW04/Dima_Yakutovich/main_methods.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    if block_given?
+
+      my_each { |i| output << yield(i) }
+    else to_enum
+    end
+    output
+  end
+  def my_select
+    if block_given?
```
Comment (B0RlS) [inline HW04/Dima_Yakutovich/main_methods.rb:28]:
```suggestion
    output
  end
  
  def my_select
    if block_given?
```

File: HW04/Dima_Yakutovich/spec/method_tests.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  end
+end
+
+RSpec.describe Array do
+  subject { Array.new([1, 2, 3, 4, 5, 256]) }
+  describe '#my_each' do
+    context 'when def each' do
+      it 'should return column of output' do
+        expectation = expect do
+          (subject.my_each { |i| puts i * 2 })
```
Comment (B0RlS) [inline HW04/Dima_Yakutovich/spec/method_tests.rb:34]: You can write "RSpec.describe Array do" once and select 3 contexts in it to test your methods

File: HW04/Dima_Yakutovich/spec/method_tests.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+require_relative '../main_methods.rb'
+require 'rspec'
+
```
Comment (B0RlS) [inline HW04/Dima_Yakutovich/spec/method_tests.rb:3]: You are testing your Array class. You should change the name of the test file to _main_methods_spec.rb_

File: HW04/README.md [hunk 4]
```diff
@@ -1,18 +1 @@
...
-### Tasks:
-
-- Implement your own `Array#my_map`, `Array#my_select`, `Array#my_each`.
-- Write specs for each of your methods usign RSpec.
-
-Notes: implement these methods in your custom module and then include it in the `Array` class.
-Important: Do not use #each, #select, #map in your implementation.
-
-
+## HW04
```
Comment (vladhilko) [inline HW04/README.md:1]: Could you return back these changes? 🤔

**Comment by kjkkuk** [comment]
it was amazing experience

**Comment by kjkkuk** [comment]
upd with rubocop-rspec

File: HW04/Dima_Yakutovich/array_methods.rb [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+# comment
+class Array
+  def my_each
+    if block_given?
+      i = 0
+      while i < length
+        yield (self[i])
+        i += 1
+      end
+    else to_enum
```
Comment (vladhilko) [inline HW04/Dima_Yakutovich/array_methods.rb:12]: Do you check this case in the spec? 🤔

File: HW04/Dima_Yakutovich/array_methods.rb [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    end
+  end
+
+  def my_map
+    output = []
+    if block_given?
+      my_each { |i| output << yield(i) }
+    else to_enum
+    end
```
Comment (vladhilko) [inline HW04/Dima_Yakutovich/array_methods.rb:21]: We don't use this style 😬
```suggestion
    else
       to_enum
    end
```

File: HW04/Dima_Yakutovich/spec/tests_array_spec.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    context 'when default map' do
+      it 'returns upd array' do
+        expect(data.my_map { |i| i * 2 })
+          .to eq [2, 4, 6, 8, 10, 512]
+      end
+    end
+  end
+
+  describe '#my_select' do
+    context 'when default select' do
```
Comment (vladhilko) [inline HW04/Dima_Yakutovich/spec/tests_array_spec.rb:19]: What does your description mean? 🤔 We're using context if we want to separate specs based on conditions ( `empty/not empty array, with/without block`)

File: HW04/Dima_Yakutovich/spec/tests_array_spec.rb [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+      it 'returns instance of integer into array with select' do
+        expect(data.my_select { |x| x.instance_of?(Integer) })
+          .to eq [1, 2, 3, 4, 5, 256]
+      end
+    end
+  end
+
+  describe '#my_each' do
+    context 'when def each' do
```
Comment (vladhilko) [inline HW04/Dima_Yakutovich/spec/tests_array_spec.rb:28]: Do you have a cave when an array is empty? 🤔 Looks like this case doesn't work with your implementation 👀

**Pull request merged**